### PR TITLE
/validator/attestations endpoint for rest api

### PR DESCRIPTION
**Pull request by rolfyone**

### CODE REVIEW

File: data/provider/src/main/java/tech/pegasys/artemis/api/ChainDataProvider.java [hunk 1]
```diff
@@ -139,6 +146,36 @@ CombinedChainDataClient getCombinedChainDataClient() {
...
+    if (block.isEmpty()) {
+      return Optional.empty();
+    }
+
+    tech.pegasys.artemis.datastructures.state.BeaconState state =
+        chainStorageClient.getBestBlockRootState();
+    int committeeCount = get_committee_count_at_slot(state, slot).intValue();
+    if (committeeIndex < 0 || committeeIndex > committeeCount) {
+      throw new IllegalArgumentException(
+          "Invalid committee index provided - expected between 0 and " + (committeeCount - 1));
```
Comment (macfarla) [inline data/provider/src/main/java/tech/pegasys/artemis/api/ChainDataProvider.java:168]: what happens if it's == committeeCount? This msg doesn't quite match the if statement

File: data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/validatorhandlers/AttestationHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+    when(context.queryParamMap()).thenReturn(params);
+    when(provider.isStoreAvailable()).thenReturn(false);
+    handler.handle(context);
+
+    verify(provider).isStoreAvailable();
+  }
+
+  @Test
+  void shouldHaveAttestationReturned() throws Exception {
```
Comment (macfarla) [inline data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/validatorhandlers/AttestationHandlerTest.java]:
```suggestion
  void shouldReturnAttestation() throws Exception {
```

File: data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/validatorhandlers/AttestationHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    when(provider.getUnsignedAttestationAtSlot(UnsignedLong.ONE, 1))
+        .thenReturn(Optional.of(attestation));
+    handler.handle(context);
+
+    verify(provider).isStoreAvailable();
+    verify(context).result(jsonProvider.objectToJSON(attestation));
+  }
+
+  @Test
+  void shouldHaveNoAttestationReturnedIfNotFound() throws Exception {
```
Comment (macfarla) [inline data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/validatorhandlers/AttestationHandlerTest.java]:
```suggestion
  void shouldReturnNoAttestationIfNotFound() throws Exception {
```

File: data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/validatorhandlers/AttestationHandler.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            name = SLOT,
+            description = "The slot to create the attestation for (uint64)",
+            required = true),
+        @OpenApiParam(
+            name = COMMITTEE_INDEX,
+            description = "The index of the committee making the attestation.",
+            required = true)
+      },
+      description =
+          "Request that the node return the beacon chain block that matches the provided criteria.",
```
Comment (macfarla) [inline data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/validatorhandlers/AttestationHandler.java]: this ain't right

File: data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/validatorhandlers/AttestationHandler.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      responses = {
+        @OpenApiResponse(
+            status = RES_OK,
+            content = @OpenApiContent(from = Attestation.class),
+            description = "Returns an attestation with a default signature."),
+        @OpenApiResponse(status = RES_BAD_REQUEST, description = "Invalid parameter supplied"),
+        @OpenApiResponse(status = RES_NOT_FOUND, description = "Specified block not found")
```
Comment (macfarla) [inline data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/validatorhandlers/AttestationHandler.java]: slot not found?

File: data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/validatorhandlers/AttestationHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+  private final JsonProvider jsonProvider = new JsonProvider();
+  private AttestationHandler handler;
+  private Attestation attestation = new Attestation(DataStructureUtil.randomAttestation(1111));
+
+  @BeforeEach
+  public void setup() {
+    handler = new AttestationHandler(provider, jsonProvider);
+  }
+
+  private void badRequestParamsTest(final Map<String, List<String>> params, String message)
```
Comment (macfarla) [inline data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/validatorhandlers/AttestationHandlerTest.java]: convention but we normally put non-test util methods at the bottom after all the tests

File: data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/validatorhandlers/AttestationHandler.java (hunk 4 above)
Comment (rolfyone) [reply data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/validatorhandlers/AttestationHandler.java]: this is copy/paste and not update :/

File: data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/validatorhandlers/AttestationHandlerTest.java (hunk 3 above)
Comment (macfarla) [reply data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/validatorhandlers/AttestationHandlerTest.java]: or shouldReturnNotFoundIfNoAttestationFound ?

File: data/provider/src/main/java/tech/pegasys/artemis/api/ChainDataProvider.java (hunk 1 above)
Comment (rolfyone) [reply data/provider/src/main/java/tech/pegasys/artemis/api/ChainDataProvider.java:168]: i'll fix this, but i'll also look at an upper boundary test, it looks like its not done if that check is wrong like that.

File: data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/validatorhandlers/AttestationHandlerTest.java (hunk 6 above)
Comment (rolfyone) [reply data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/validatorhandlers/AttestationHandlerTest.java]: this is the actual test method all the following tests use... i can move it if you'd prefer...

File: data/provider/src/test/java/tech/pegasys/artemis/api/ChainDataProviderTest.java [hunk 7]
```diff
@@ -315,4 +318,78 @@ void getStateByBlockRoot_shouldReturnBeaconStateWhenFound()
...
+  @Test
+  void getUnsignedAttestationAtSlot_shouldReturnEmptyIfCommitteeAboveRange() {
+    ChainDataProvider provider =
+        new ChainDataProvider(chainStorageClient, mockCombinedChainDataClient);
+    when(mockCombinedChainDataClient.isStoreAvailable()).thenReturn(true);
+    when(mockCombinedChainDataClient.isFinalized(ZERO)).thenReturn(false);
+    assertThrows(
+        IllegalArgumentException.class, () -> provider.getUnsignedAttestationAtSlot(ZERO, 1));
+    verify(mockCombinedChainDataClient).isStoreAvailable();
+    verify(mockCombinedChainDataClient).isFinalized(ZERO);
```
Comment (macfarla) [inline data/provider/src/test/java/tech/pegasys/artemis/api/ChainDataProviderTest.java]: these 3 tests are almost the same - wonder if we can refactor to make the differences clear?

Comment (rolfyone) [reply data/provider/src/test/java/tech/pegasys/artemis/api/ChainDataProviderTest.java]: refactored.

File: data/provider/src/test/java/tech/pegasys/artemis/api/ChainDataProviderTest.java [hunk 8]
```diff
@@ -315,4 +318,70 @@ void getStateByBlockRoot_shouldReturnBeaconStateWhenFound()
...
+    when(mockCombinedChainDataClient.isStoreAvailable()).thenReturn(false);
+    Optional<Attestation> optional = provider.getUnsignedAttestationAtSlot(ZERO, 0);
+    verify(mockCombinedChainDataClient).isStoreAvailable();
+    assertTrue(optional.isEmpty());
+  }
+
+  @Test
+  void getUnsignedAttestationAtSlot_shouldReturnEmptyIfSlotIsFinalized() {
+    getUnsignedAttestationAtSlot_throwsIllegalArgumentException(0, true);
+  }
```
Comment (macfarla) [inline data/provider/src/test/java/tech/pegasys/artemis/api/ChainDataProviderTest.java:335]: nice

**Review by macfarla — APPROVED** [review]

**Pull request merged**